internal/audits/audits.go

```diff
@@ -0,0 +1,105 @@
+package audits
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/opdev/opcap/internal/operator"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// AuditOption is the function type for passing an option to an audit
+type AuditOption func(options *options) error
+
+// WithSubscription adds a subscription object to the audit
+func WithSubscription(subscription *operator.SubscriptionData) AuditOption {
+	return func(options *options) error {
+		if subscription == nil {
+			return fmt.Errorf("subscription data cannot be nil")
+		}
+		options.Subscription = subscription
+		return nil
+	}
+}
+
+// WithOperatorGroupData adds an operatorgroupdata objec to the audit
+func WithOperatorGroupData(operatorGroupData *operator.OperatorGroupData) AuditOption {
+	return func(options *options) error {
+		if operatorGroupData == nil {
+			return fmt.Errorf("operatorgroupdata cannot be nil")
+		}
+		options.operatorGroupData = operatorGroupData
+		return nil
+	}
+}
+
+// WithNamespace adds a namespace for the audit
+func WithNamespace(namespace string) AuditOption {
+	return func(options *options) error {
+		if namespace == "" {
+			return fmt.Errorf("namespace cannot be empty")
+		}
+		options.namespace = namespace
+		return nil
+	}
+}
+
+// WithClient adds a client to the audit
+func WithClient(client operator.Client) AuditOption {
+	return func(options *options) error {
+		if client == nil {
+			return fmt.Errorf("client cannot be nil")
+		}
+		options.client = client
+		return nil
+	}
+}
+
+// WithTimeout adds a timeout duration to the audit
+func WithTimeout(csvWaitTime int) AuditOption {
+	return func(options *options) error {
+		options.csvWaitTime = time.Duration(csvWaitTime)
+		return nil
+	}
+}
+
+// WithOcpVersion adds the OCP version to the audit
+func WithOcpVersion(ocpVersion string) AuditOption {
+	return func(options *options) error {
+		options.OcpVersion = ocpVersion
+		return nil
+	}
+}
+
+type options struct {
```

Comment:
Here we have something that is not only options. Yes it carries options but also stores data to process reports such as complete captured CSVs and complete rendered operands by the end of each complete audit hence the name capAudit. With that data we intend to write and store on a database at each single audit for each operator package/channel/InstallMode. So this data should be written before passing to the next operator. We can change that design and separate those pieces of data and have options and something like auditData. Otherwise I think calling this options may confuse people since this struct will be storing data as well.

Reply:
I see what you mean. I think what you're saying here exactly shows the point that these "audits" are doing way too much. They should not be handling the reporting. This might be something that we need to tackle in another PR.

Reply:
Exactly. The two previous comments are related and yes. It's definitely doing too much. Not sure how to approach the separation of concerns here but it needs it. It can totally be another PR or PRs if needed.

Reply:
Sounds good.